flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/OnlineKMeans.java

```diff
@@ -0,0 +1,407 @@
+
+        IterationBody body =
+                new OnlineKMeansIterationBody(
+                        DistanceMeasure.getInstance(getDistanceMeasure()),
+                        getK(),
+                        getDecayFactor(),
+                        getGlobalBatchSize());
+
+        DataStream<KMeansModelData> onlineModelData =
+                Iterations.iterateUnboundedStreams(
+                                DataStreamList.of(initModelData), DataStreamList.of(points), body)
+                        .get(0);
+
+        Table onlineModelDataTable = tEnv.fromDataStream(onlineModelData);
+        OnlineKMeansModel model = new OnlineKMeansModel().setModelData(onlineModelDataTable);
+        ReadWriteUtils.updateExistingParams(model, paramMap);
+        return model;
+    }
+
+    /** Saves the metadata AND bounded model data table (if exists) to the given path. */
+    @Override
+    public void save(String path) throws IOException {
+        if (initModelDataTable != null) {
+            ReadWriteUtils.saveModelData(
+                    KMeansModelData.getModelDataStream(initModelDataTable),
+                    path,
+                    new KMeansModelData.ModelDataEncoder());
+        }
+
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static OnlineKMeans load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        OnlineKMeans onlineKMeans = ReadWriteUtils.loadStageParam(path);
+
+        String initModelDataPath = ReadWriteUtils.getDataPath(path);
+        if (Files.exists(Paths.get(initModelDataPath))) {
+            StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
+
+            DataStream<KMeansModelData> initModelDataStream =
+                    ReadWriteUtils.loadModelData(env, path, new KMeansModelData.ModelDataDecoder());
+
+            onlineKMeans.initModelDataTable = tEnv.fromDataStream(initModelDataStream);
+        }
+
+        return onlineKMeans;
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    private static class OnlineKMeansIterationBody implements IterationBody {
+        private final DistanceMeasure distanceMeasure;
+        private final int k;
+        private final double decayFactor;
+        private final int batchSize;
+
+        public OnlineKMeansIterationBody(
+                DistanceMeasure distanceMeasure, int k, double decayFactor, int batchSize) {
+            this.distanceMeasure = distanceMeasure;
+            this.k = k;
+            this.decayFactor = decayFactor;
+            this.batchSize = batchSize;
+        }
+
+        @Override
+        public IterationBodyResult process(
+                DataStreamList variableStreams, DataStreamList dataStreams) {
+            DataStream<KMeansModelData> modelData = variableStreams.get(0);
+            DataStream<DenseVector> points = dataStreams.get(0);
+
+            int parallelism = points.getParallelism();
```

Comment:
Would it be useful to explicitly check and throw exception if `parallelism <= batchSize`?

Reply:
I guess what you want to express is to make sure `parallelism <= batchSize`, and to throw exception if `parallelism > batchSize`. I cannot think of a valid reason to do this check.

I think even if the number of elements in a batch is smaller than parallelism, the training process should still function correctly, because when batches are created according to time window, the operator would have no idea about how many elements are in each batch.

If the case above is correct, the purpose if doing the `parallelism <= batchSize`, as I can think of, is performance issues. But I think the same issue also exists when `batchSize` is just a bit larger than `parallelism`, that each subtask is only allocated with 1 or 2 training data. In this situation, a strict `parallelism <= batchSize` threshold won't solve the performance problem.

Reply:
Yes, I think the reason is related to performance.

If `parallelism > batchSize`,  it effectively means some slot (with its CPU resource) it definitely wasted. Is there any reason user would want to do this? If not, it means user must have chosen this setup by mistake. Would it be more user friendly to alert user of this issue by throwing an exception?

Reply:
OK. I'll add the check and relevant test case.

Reply:
I hold a different opinion on this. 
We probably should also support when the paralleism is smaller than batch size for consistent user experience. But we could allert the user by a warning.

Reply:
@zhipeng93 Do you think there is any reason that user would want to actually run OnlinkeKMeans with parallelism < batchSize, and if so, could you provide some detail?

And if no, could you explain why we should actually support this?

Reply:
As pointed out by @yunfengzhou-hub , a use case could be:
```I think even if the number of elements in a batch is smaller than parallelism, the training process should still function correctly, because when batches are created according to time window, the operator would have no idea about how many elements are in each batch.```

By `consistent user experience`, I mean that given a input of training set, users should be able to run it on one worker or arbitrary number of workers.

Reply:
Thanks for the discussion @zhipeng93.

To summarize our offline discussion, we agreed that there is no reasonable reason that users would want to run the training process with `batchStrategy=size` and `parallelism > batchSize`. We will throw exception here to prevent users from wasting resources. And we will not throw this exception if the algorithm is configured with `batchStrategy=time`.